**Classificação página 83**

In [1]:
import time
from sklearn.datasets import fetch_openml

In [12]:
ti = time.time()

mnist_str = fetch_openml('mnist_784', version=1)

tf = time.time()
print('tempo download:', tf-ti, 'segundos')

tempo download: 3.7123024463653564 segundos


In [3]:
import pickle

In [5]:
mnist_bunch = ""
with open('mnist.pkl', 'wb') as bunch:
    pickle.dump(mnist_bunch, bunch, protocol=pickle.HIGHEST_PROTOCOL)

In [6]:
print(bunch, '\n\nmnist.pkl está aberto e pronto para receber dados em formato binário')

<_io.BufferedWriter name='mnist.pkl'> 

mnist.pkl está aberto e pronto para receber dados em formato binário


In [7]:
print(type(bunch), '\n\nbunch aponta para um objeto da classe _io.BufferedReader\nA referência bunch lê arquivos em modo binário')

<class '_io.BufferedWriter'> 

bunch aponta para um objeto da classe _io.BufferedReader
A referência bunch lê arquivos em modo binário


In [8]:
with open('mnist.pkl', 'rb') as bunch: 
    mnist_str = pickle.load(bunch)
# bunch aponta para um objeto da classe _io.BufferedWriter.
# A referência bunch escreve  m arquivos em modo binário

In [13]:
type(mnist_str)

sklearn.utils._bunch.Bunch

In [14]:
mnist_str.keys()

dict_keys(['data', 'target', 'frame', 'categories', 'feature_names', 'target_names', 'DESCR', 'details', 'url'])

In [15]:
mnist_str['target']

0        5
1        0
2        4
3        1
4        9
        ..
69995    2
69996    3
69997    4
69998    5
69999    6
Name: class, Length: 70000, dtype: category
Categories (10, object): ['0', '1', '2', '3', ..., '6', '7', '8', '9']

In [16]:
print(mnist_str['feature_names'])

['pixel1', 'pixel2', 'pixel3', 'pixel4', 'pixel5', 'pixel6', 'pixel7', 'pixel8', 'pixel9', 'pixel10', 'pixel11', 'pixel12', 'pixel13', 'pixel14', 'pixel15', 'pixel16', 'pixel17', 'pixel18', 'pixel19', 'pixel20', 'pixel21', 'pixel22', 'pixel23', 'pixel24', 'pixel25', 'pixel26', 'pixel27', 'pixel28', 'pixel29', 'pixel30', 'pixel31', 'pixel32', 'pixel33', 'pixel34', 'pixel35', 'pixel36', 'pixel37', 'pixel38', 'pixel39', 'pixel40', 'pixel41', 'pixel42', 'pixel43', 'pixel44', 'pixel45', 'pixel46', 'pixel47', 'pixel48', 'pixel49', 'pixel50', 'pixel51', 'pixel52', 'pixel53', 'pixel54', 'pixel55', 'pixel56', 'pixel57', 'pixel58', 'pixel59', 'pixel60', 'pixel61', 'pixel62', 'pixel63', 'pixel64', 'pixel65', 'pixel66', 'pixel67', 'pixel68', 'pixel69', 'pixel70', 'pixel71', 'pixel72', 'pixel73', 'pixel74', 'pixel75', 'pixel76', 'pixel77', 'pixel78', 'pixel79', 'pixel80', 'pixel81', 'pixel82', 'pixel83', 'pixel84', 'pixel85', 'pixel86', 'pixel87', 'pixel88', 'pixel89', 'pixel90', 'pixel91', 'pixel9

In [17]:
mnist_str['data']

pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
0           0       0       0       0       0       0       0       0       0   
1           0       0       0       0       0       0       0       0       0   
2           0       0       0       0       0       0       0       0       0   
3           0       0       0       0       0       0       0       0       0   
4           0       0       0       0       0       0       0       0       0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
69995       0       0       0       0       0       0       0       0       0   
69996       0       0       0       0       0       0       0       0       0   
69997       0       0       0       0       0       0       0       0       0   
69998       0       0       0       0       0       0       0       0       0   
69999       0       0       0       0       0       0       0       0       0   

       pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  \
0            0  ...         0         0         0         0         0   
1            0  ...         0         0         0         0         0   
2            0  ...         0         0         0         0         0   
3            0  ...         0         0         0         0         0   
4            0  ...         0         0         0         0         0   
...        ...  ...       ...       ...       ...       ...       ...   
69995        0  ...         0         0         0         0         0   
69996        0  ...         0         0         0         0         0   
69997        0  ...         0         0         0         0         0   
69998        0  ...         0         0         0         0         0   
69999        0  ...         0         0         0         0         0   

       pixel780  pixel781  pixel782  pixel783  pixel784  
0             0         0         0         0         0  
1             0         0         0         0         0  
2             0         0         0         0         0  
3             0         0         0         0         0  
4             0         0         0         0         0  
...         ...       ...       ...       ...       ...  
69995         0         0         0         0         0  
69996         0         0         0         0         0  
69997         0         0         0         0         0  
69998         0         0         0         0         0  
69999         0         0         0         0         0  

[70000 rows x 784 columns]

In [18]:
X, y = mnist_str['data'], mnist_str['target']

In [19]:
print(type(X), type(y))

<class 'pandas.core.frame.DataFrame'> <class 'pandas.core.series.Series'>


**conveter de str para int**<br>

In [20]:
y = y.astype(int)

In [21]:
type(y), y.iloc[9], y.iloc[9].shape

(pandas.core.series.Series, np.int64(4), ())

In [22]:
type(X), X.iloc[9], X.iloc[9].shape

(pandas.core.frame.DataFrame,
 pixel1      0
 pixel2      0
 pixel3      0
 pixel4      0
 pixel5      0
            ..
 pixel780    0
 pixel781    0
 pixel782    0
 pixel783    0
 pixel784    0
 Name: 9, Length: 784, dtype: int64,
 (784,))

In [23]:
X.iloc[9].values.reshape(28,28)

array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0, 189, 190,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0, 143, 247, 153,   0,   0,
          0,   0],
       [  

In [24]:
X.iloc[7].values.reshape(28,28)

array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  38,  43,
        105, 255, 253, 253, 253, 253, 253, 174,   6,   0,   0,   0,   0,
          0,   0],
       [  

In [25]:
import numpy as np

In [26]:
np.set_printoptions(linewidth=200)

X.iloc[9].values.reshape(28,28)

array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 189, 190,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 143, 247, 153,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0

In [27]:
X.iloc[7].values.reshape(28,28)

array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  38,  43, 105, 255, 253, 253, 253, 253, 253, 174,   6,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,  43, 139, 224, 226, 252, 253, 252, 252, 252, 252, 252

<br>**Há 70 mil imagens e cada imagem possui 784 características (cada imagem tem 784=28×28 pixels).**<br>
**Cada característica representa a intensidade de um pixel, de 0 (branco) a 255 (preto).**<br>
**Vamos visualizar um dígito com a função imshow() do Matplotlib.**<br>
**Só precisa pegar um vetor de features de uma instância e remodelá-lo para um array de 28 × 28.**<br><br>

In [ ]:
y.iloc[7]

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
dig = X.iloc[7]
img = dig.values.reshape(28,28)
plt.imshow(img, cmap=plt.cm.gray_r)
#plt.imshow(img)
plt.axis('off');

In [ ]:
plt.figure(figsize=(5, 3))
for i in range(50):    
    plt.subplot(5,10, i + 1)
    plt.imshow(X.iloc[i].values.reshape(28,28), cmap='binary')               
    plt.axis('off')

**1 - Separar dados de teste**<br>
**O dataset MNIST disponível no OpenML segue a divisão tradicional de 60.000 imagens para treinamento e 10.000 para teste. Isto é consistente com a versão original do dataset, que foi projetado para facilitar a criação e avaliação de modelos de aprendizado de máquina, especialmente em tarefas de classificação de dígitos manuscritos.**<br>

In [28]:
X_train, X_test, y_train, y_test = X[:60000], X[60000:], y[:60000], y[60000:]

In [29]:
y_train[59999]

np.int64(8)

In [30]:
y_test[60000]

np.int64(7)

In [31]:
y_train.shape, y_test.shape

((60000,), (10000,))

**2 - Embaralhar os dados para**<br>
* Evitar dependências estruturais ou padrões não intencionais (por exemplo, todos os dígitos 0 seguidos de todos os dígitos 1, etc.)
* Reduzir o risco de overfitting (o modelo pode começar a "memorizar" a sequência de dados em vez de aprender a generalizar)
* Preparar para cross-validation, na qual é essencial que os dados sejam embaralhados para garantir que as divisões dos dados sejam aleatórias.
* Em modelo como Stochastic Gradient Descent (SGD), embaralhar os dados em cada época pode ajudar a reduzir a correlação entre as amostras consecutivas<br>

In [32]:
from sklearn.utils import shuffle

In [33]:
X_train, y_train = shuffle(X_train, y_train, random_state=42)

12628    False
37730    False
39991    False
8525     False
8279     False
         ...  
54343    False
38158    False
860      False
15795    False
56422    False
Name: class, Length: 60000, dtype: bool

**<center>Parou aqui em 11 de outubro**

**<center>Treinando classificador binário**

**Classificador binário capaz de distinguir apenas entre duas classes: 5 e não 5 (livro Géron)**

**Classificador binário capaz de distinguir apenas entre duas classes: 1 e 7**

In [ ]:
#                 Previsto 1               Previsto 7
# Real 1     previu 1 e é 1 mesmo       previu 7 mas é 1
# Real 7      previu 1 mas é 7       previu 7 e é 7 mesmo


#                          Previsto
#                    Positivo    Negativo
#         Positivo     TP          FN
#   Real
#         Negativo     FP          TN 

**Matriz de confusão é uma tabela usada em problemas de classificação para avaliar o desempenho de um modelo, comparando as previsões do modelo
com os valores reais dos dados.**<br>

**Ela mostra quantas das classificações corretas e incorretas foram feitas pelo modelo.**<br>

**Os elementos da diagonal principal representam as previsões corretas e os elementos da diagonal secundária representam as previsões incorretas.**<br>

**<center>Avaliar o modelo com cross_val_predict**

**A função cross_val_score retorna as pontuações (scores) do modelo após cada rodada de validação cruzada.**<br>
forest_scores = cross_val_score(forest_reg,train_preds,train_labels,coring='neg_mean_squared_error',cv=10)<br>

**A função cross_val_predict retorna as predições feitas durante a validação cruzada.**

**A matriz de confusão fornece muitas informações, mas às vezes é interessante usar uma métrica mais concisa.**<br>
**Uma coisa interessante a ser observada é a <u>acurácia das previsões positivas</u> que é chamada de <u>precisão</u> do classificador.**<br>

$$ \text{Precisão} = \frac{TP}{TP + FP} $$

In [ ]:
#   Exemplo
#   Supondo 200 amostras com 102 números 1 e 98 números 7
#   96     6
#    0     98

#   TP     FN
#   FP     TN 

**Precisão foca no FP. Deu 100% porque toda vez que ele disse que é 1, ele acertou.**<br>
**Mas ele falhou em reconhecer 6 números 1**<br>
**A precisão sozinha não reflete totalmente o desempenho do modelo. Por isso, é utilizada em conjunto com outra métrica chamada recall (revocação), também conhecida como sensitivity (sensibilidade) ou taxa de verdadeiros positivos**<br>
$$ \text{Recall} = \frac{TP}{TP + FN} $$

**A precisão foca no FP: Nenhuma previsão do número 1 foi errada, então a precisão foi 100%.**<br>
**Recall foca no FN: O modelo não conseguiu identificar todos os 1 (errou 7 deles), então o recall foi 93.13%.**<br>
Recall se refere à capacidade do modelo de "recuperar" (recall) todas as instâncias positivas, no nosso exemplo, o número 1.

Géron pg 91
Muitas vezes, é conveniente combinar precisão e revocação em uma única métrica chamada
pontuação F1, principalmente se você precisar de uma forma simples de comparar dois
classificadores. A pontuação F1 é a média harmônica da precisão e revocação (Equação
3-3). Enquanto a média regular trata igualmente todos os valores, a média harmônica
dá muito mais peso aos valores mais baixos. Como resultado, o classificador só obterá
uma pontuação F1 alta se a revocação e a precisão forem altas.

$$\text{F}_1 = \frac{1}{2} \left( \frac{TP}{TP + FN} + \frac{TN}{TN + FP} \right)$$